Dockerfile

```diff
@@ -1,69 +1,96 @@
 # Pelican Production Dockerfile
 
 # ================================
-# Stage 1: Build PHP Dependencies
+# Stage 1: Build PHP Base Image
 # ================================
-FROM --platform=$TARGETOS/$TARGETARCH php:8.3-fpm-alpine AS composer
+FROM --platform=$TARGETOS/$TARGETARCH php:8.3-fpm-alpine AS base
+
+ADD --chmod=0755 https://github.com/mlocati/docker-php-extension-installer/releases/latest/download/install-php-extensions /usr/local/bin/
+
+RUN install-php-extensions bcmath gd intl zip opcache pcntl posix pdo_mysql
+
+RUN rm /usr/local/bin/install-php-extensions
+
+# ================================
+# Stage 2-1: Composer Install
+# ================================
+FROM --platform=$TARGETOS/$TARGETARCH base AS composer
 
 WORKDIR /build
 
-COPY . ./
 COPY --from=composer:latest /usr/bin/composer /usr/local/bin/composer
 
-# Install required libraries and PHP extensions
-RUN apk update && apk add --no-cache \
-    libpng-dev libjpeg-turbo-dev freetype-dev libzip-dev icu-dev \
-    zip unzip curl \
-    && docker-php-ext-install bcmath gd intl zip opcache pcntl posix pdo_mysql
+# Copy bare minimum to install Composer dependencies
+COPY composer.json composer.lock ./
 
-RUN composer install --no-dev --optimize-autoloader
+RUN composer install --no-dev --no-interaction --no-autoloader --no-scripts
 
 # ================================
-# Stage 2: Build Frontend Assets
+# Stage 2-2: Yarn Install
 # ================================
 FROM --platform=$TARGETOS/$TARGETARCH node:20-alpine AS yarn
 
 WORKDIR /build
 
-COPY --from=composer /build .
+# Copy bare minimum to install Yarn dependencies
+COPY package.json yarn.lock ./
 
 RUN yarn config set network-timeout 300000 \
-    && yarn install --frozen-lockfile \
-    && yarn run build
+    && yarn install --frozen-lockfile
 
 # ================================
-# Stage 3: Build Final Application Image
+# Stage 3-1: Composer Optimize
 # ================================
-FROM --platform=$TARGETOS/$TARGETARCH php:8.3-fpm-alpine
+FROM --platform=$TARGETOS/$TARGETARCH composer AS composerbuild
 
-WORKDIR /var/www/html
+# Copy full code to optimize autoload
+COPY --exclude=Caddyfile --exclude=docker/ . ./
 
-COPY --from=composer:latest /usr/bin/composer /usr/local/bin/composer
+RUN composer dump-autoload --optimize
+
+# ================================
+# Stage 3-2: Build Frontend Assets
+# ================================
+FROM --platform=$TARGETOS/$TARGETARCH yarn AS yarnbuild
+
+WORKDIR /build
+
+# Copy full code
+COPY --exclude=Caddyfile --exclude=docker/ . ./
+COPY --from=composer /build .
+
+RUN yarn run build
+
+# ================================
+# Stage 4: Build Final Application Image
+# ================================
+FROM --platform=$TARGETOS/$TARGETARCH base AS final
+
+WORKDIR /var/www/html
 
 # Install additional required libraries
 RUN apk update && apk add --no-cache \
-    libpng-dev libjpeg-turbo-dev freetype-dev libzip-dev icu-dev \
```

Comment:
Those packages are needed in the final container.

Reply:
They shouldn't be. The [above ext install tool](https://github.com/mlocati/docker-php-extension-installer) installs the dev packages for building, and removes them afterwards. Only the non dev packages are left behind in the final image.

Reply:
same here

Reply:
Sorry, I added it above in the original PR message to make sure it's clearer that I added that helper script to deal with this, and it's not just the same thing.